File: native-plugins/android/FaceDetectionPlugin.kt
package com.photocropper.plugins

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.util.Base64
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.face.Face
import com.google.mlkit.vision.face.FaceDetection
import com.google.mlkit.vision.face.FaceDetector
import com.google.mlkit.vision.face.FaceDetectorOptions

@CapacitorPlugin(name = "FaceDetection")
class FaceDetectionPlugin : Plugin() {

    private var detector: FaceDetector? = null

    @PluginMethod
    fun initialize(call: PluginCall) {
        try {
            val options = FaceDetectorOptions.Builder()
                .setPerformanceMode(FaceDetectorOptions.PERFORMANCE_MODE_FAST)
                .setLandmarkMode(FaceDetectorOptions.LANDMARK_MODE_NONE)
                .setClassificationMode(FaceDetectorOptions.CLASSIFICATION_MODE_NONE)
                .setMinFaceSize(0.15f)
                .build()

            detector = FaceDetection.getClient(options)
            
            val result = JSObject()
            result.put("success", true)
            call.resolve(result)
        } catch (e: Exception) {
            val result = JSObject()
            result.put("success", false)
            call.resolve(result)
        }
    }

    @PluginMethod
    fun isAvailable(call: PluginCall) {
        val result = JSObject()
        result.put("available", true)
        result.put("platform", "android")
        call.resolve(result)
    }

    @PluginMethod
    fun dispose(call: PluginCall) {
        detector?.close()
        detector = null
        call.resolve()
    }

    @PluginMethod
    fun detectFace(call: PluginCall) {
        val imageData = call.getString("imageData")
        if (imageData == null) {
            call.reject("Missing imageData parameter")
            return
        }

        val ovalBoundsObj = call.getObject("ovalBounds")
        if (ovalBoundsObj == null) {
            call.reject("Missing ovalBounds parameter")
            return
        }

        val ovalBounds = OvalBounds(
            centerX = ovalBoundsObj.getDouble("centerX"),
            centerY = ovalBoundsObj.getDouble("centerY"),
            radiusX = ovalBoundsObj.getDouble("radiusX"),
            radiusY = ovalBoundsObj.getDouble("radiusY")
        )

        val bitmap = decodeBase64Image(imageData)
        if (bitmap == null) {
            call.reject("Failed to decode image")
            return
        }

        // Initialize detector if needed
        if (detector == null) {
            val options = FaceDetectorOptions.Builder()
                .setPerformanceMode(FaceDetectorOptions.PERFORMANCE_MODE_FAST)
                .setMinFaceSize(0.15f)
                .build()
            detector = FaceDetection.getClient(options)
        }

        val inputImage = InputImage.fromBitmap(bitmap, 0)

        detector?.process(inputImage)
            ?.addOnSuccessListener { faces ->
                val result = processFaces(faces, bitmap.width, bitmap.height, ovalBounds)
                bitmap.recycle()
                call.resolve(result)
            }
            ?.addOnFailureListener { e ->
                bitmap.recycle()
                val result = JSObject()
                result.put("faceDetected", false)
                result.put("guidance", "no_face")
                call.resolve(result)
            }
    }

    private fun processFaces(faces: List<Face>, imageWidth: Int, imageHeight: Int, ovalBounds: OvalBounds): JSObject {
        val result = JSObject()

        if (faces.isEmpty()) {
            result.put("faceDetected", false)
            result.put("guidance", "no_face")
            return result
        }

        if (faces.size > 1) {
            result.put("faceDetected", true)
            result.put("guidance", "multiple_faces")
            return result
        }

        val face = faces[0]
        val boundingBox = face.boundingBox

        // Normalize bounds to 0-1 range
        val bounds = JSObject()
        bounds.put("x", boundingBox.left.toDouble() / imageWidth)
        bounds.put("y", boundingBox.top.toDouble() / imageHeight)
        bounds.put("width", boundingBox.width().toDouble() / imageWidth)
        bounds.put("height", boundingBox.height().toDouble() / imageHeight)

        val guidance = calculateGuidance(
            faceX = bounds.getDouble("x"),
            faceY = bounds.getDouble("y"),
            faceWidth = bounds.getDouble("width"),
            faceHeight = bounds.getDouble("height"),
            ovalBounds = ovalBounds
        )

        result.put("faceDetected", true)
        result.put("bounds", bounds)
        result.put("guidance", guidance)
        result.put("confidence", 0.9) // ML Kit doesn't provide confidence for bounding box

        return result
    }

    private fun calculateGuidance(
        faceX: Double,
        faceY: Double,
        faceWidth: Double,
        faceHeight: Double,
        ovalBounds: OvalBounds
    ): String {
        val faceCenterX = faceX + faceWidth / 2
        val faceCenterY = faceY + faceHeight / 2

        val positionThreshold = 0.08
        val faceSize = maxOf(faceWidth, faceHeight)
        val ovalSize = maxOf(ovalBounds.radiusX * 2, ovalBounds.radiusY * 2)
        val sizeRatio = faceSize / ovalSize

        // Too small = move closer
        if (sizeRatio < 0.5) {
            return "move_closer"
        }

        // Too large = move back
        if (sizeRatio > 1.1) {
            return "move_back"
        }

        // Check horizontal position
        val xOffset = faceCenterX - ovalBounds.centerX
        if (xOffset < -positionThreshold) {
            return "move_right"
        }
        if (xOffset > positionThreshold) {
            return "move_left"
        }

        // Check vertical position
        val yOffset = faceCenterY - ovalBounds.centerY
        if (yOffset < -positionThreshold) {
            return "move_down"
        }
        if (yOffset > positionThreshold) {
            return "move_up"
        }

        return "hold_still"
    }

    private fun decodeBase64Image(dataUrl: String): Bitmap? {
        return try {
            val base64String = if (dataUrl.contains(",")) {
                dataUrl.substringAfter(",")
            } else {
                dataUrl
            }
            val bytes = Base64.decode(base64String, Base64.DEFAULT)
            BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
        } catch (e: Exception) {
            null
        }
    }

    private data class OvalBounds(
        val centerX: Double,
        val centerY: Double,
        val radiusX: Double,
        val radiusY: Double
    )
}

File: native-plugins/ios/FaceDetectionPlugin.swift
import Foundation
import Capacitor
import Vision
import UIKit

@objc(FaceDetectionPlugin)
public class FaceDetectionPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "FaceDetectionPlugin"
    public let jsName = "FaceDetection"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "initialize", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "detectFace", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isAvailable", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "dispose", returnType: CAPPluginReturnPromise)
    ]
    
    private var isInitialized = false
    
    @objc func initialize(_ call: CAPPluginCall) {
        // Vision framework is always available on iOS 11+
        isInitialized = true
        call.resolve(["success": true])
    }
    
    @objc func isAvailable(_ call: CAPPluginCall) {
        call.resolve([
            "available": true,
            "platform": "ios"
        ])
    }
    
    @objc func dispose(_ call: CAPPluginCall) {
        isInitialized = false
        call.resolve()
    }
    
    @objc func detectFace(_ call: CAPPluginCall) {
        guard let imageData = call.getString("imageData") else {
            call.reject("Missing imageData parameter")
            return
        }
        
        guard let ovalBoundsObj = call.getObject("ovalBounds"),
              let centerX = ovalBoundsObj["centerX"] as? Double,
              let centerY = ovalBoundsObj["centerY"] as? Double,
              let radiusX = ovalBoundsObj["radiusX"] as? Double,
              let radiusY = ovalBoundsObj["radiusY"] as? Double else {
            call.reject("Missing or invalid ovalBounds parameter")
            return
        }
        
        let ovalBounds = OvalBounds(centerX: centerX, centerY: centerY, radiusX: radiusX, radiusY: radiusY)
        
        guard let image = imageFromDataUrl(imageData),
              let cgImage = image.cgImage else {
            call.reject("Failed to decode image")
            return
        }
        
        // Perform face detection on background thread
        DispatchQueue.global(qos: .userInteractive).async {
            self.performFaceDetection(cgImage: cgImage, ovalBounds: ovalBounds) { result in
                DispatchQueue.main.async {
                    call.resolve(result)
                }
            }
        }
    }
    
    private func performFaceDetection(cgImage: CGImage, ovalBounds: OvalBounds, completion: @escaping ([String: Any]) -> Void) {
        let request = VNDetectFaceRectanglesRequest { request, error in
            if let error = error {
                print("Face detection error: \(error)")
                completion([
                    "faceDetected": false,
                    "guidance": "no_face"
                ])
                return
            }
            
            guard let observations = request.results as? [VNFaceObservation] else {
                completion([
                    "faceDetected": false,
                    "guidance": "no_face"
                ])
                return
            }
            
            if observations.isEmpty {
                completion([
                    "faceDetected": false,
                    "guidance": "no_face"
                ])
                return
            }
            
            if observations.count > 1 {
                completion([
                    "faceDetected": true,
                    "guidance": "multiple_faces"
                ])
                return
            }
            
            let face = observations[0]
            
            // Vision uses normalized coordinates (0-1) with origin at bottom-left
            // Convert to top-left origin
            let bounds: [String: Double] = [
                "x": Double(face.boundingBox.origin.x),
                "y": Double(1.0 - face.boundingBox.origin.y - face.boundingBox.height),
                "width": Double(face.boundingBox.width),
                "height": Double(face.boundingBox.height)
            ]
            
            let guidance = self.calculateGuidance(
                faceX: bounds["x"]!,
                faceY: bounds["y"]!,
                faceWidth: bounds["width"]!,
                faceHeight: bounds["height"]!,
                ovalBounds: ovalBounds
            )
            
            completion([
                "faceDetected": true,
                "bounds": bounds,
                "guidance": guidance,
                "confidence": Double(face.confidence)
            ])
        }
        
        // Use accurate mode for better detection
        request.revision = VNDetectFaceRectanglesRequestRevision3
        
        let handler = VNImageRequestHandler(cgImage: cgImage, options: [:])
        do {
            try handler.perform([request])
        } catch {
            print("Failed to perform face detection: \(error)")
            completion([
                "faceDetected": false,
                "guidance": "no_face"
            ])
        }
    }
    
    private func calculateGuidance(faceX: Double, faceY: Double, faceWidth: Double, faceHeight: Double, ovalBounds: OvalBounds) -> String {
        let faceCenterX = faceX + faceWidth / 2
        let faceCenterY = faceY + faceHeight / 2
        
        let positionThreshold = 0.08
        let faceSize = max(faceWidth, faceHeight)
        let ovalSize = max(ovalBounds.radiusX * 2, ovalBounds.radiusY * 2)
        let sizeRatio = faceSize / ovalSize
        
        // Too small = move closer
        if sizeRatio < 0.5 {
            return "move_closer"
        }
        
        // Too large = move back
        if sizeRatio > 1.1 {
            return "move_back"
        }
        
        // Check horizontal position
        let xOffset = faceCenterX - ovalBounds.centerX
        if xOffset < -positionThreshold {
            return "move_right"
        }
        if xOffset > positionThreshold {
            return "move_left"
        }
        
        // Check vertical position
        let yOffset = faceCenterY - ovalBounds.centerY
        if yOffset < -positionThreshold {
            return "move_down"
        }
        if yOffset > positionThreshold {
            return "move_up"
        }
        
        return "hold_still"
    }
    
    private func imageFromDataUrl(_ dataUrl: String) -> UIImage? {
        let base64String: String
        if dataUrl.contains(",") {
            base64String = String(dataUrl.split(separator: ",").last ?? "")
        } else {
            base64String = dataUrl
        }
        
        guard let data = Data(base64Encoded: base64String) else {
            return nil
        }
        
        return UIImage(data: data)
    }
}

private struct OvalBounds {
    let centerX: Double
    let centerY: Double
    let radiusX: Double
    let radiusY: Double
}

